src/modules/job-list/match.c

```diff
@@ -0,0 +1,654 @@
+        t = job->t_submit;
+    else if (tv->t_type == MATCH_T_DEPEND) {
+        /* if submit_version < 1, it means it was not set.  This is
+         * before the introduction of event `validate` after 0.41.1.
+         * Before the introduction of this event, t_submit and
+         * t_depend are the same.
+         */
+        if (job->submit_version < 1)
+            t = job->t_submit;
+        else if (job->states_mask & FLUX_JOB_STATE_DEPEND)
+            t = job->t_depend;
+        else
+            return false;
+    }
+    else if (tv->t_type == MATCH_T_RUN
+             && (job->states_mask & FLUX_JOB_STATE_RUN))
+        t = job->t_run;
+    else if (tv->t_type == MATCH_T_CLEANUP
+             && (job->states_mask & FLUX_JOB_STATE_CLEANUP))
+        t = job->t_cleanup;
+    else if (tv->t_type == MATCH_T_INACTIVE
+             && (job->states_mask & FLUX_JOB_STATE_INACTIVE))
+        t = job->t_inactive;
+    else
+        return false;
+
+    if (tv->t_comp == MATCH_GREATER_THAN_EQUAL)
+        return t >= tv->t_value;
+    else if (tv->t_comp == MATCH_LESS_THAN_EQUAL)
+        return t <= tv->t_value;
+    else if (tv->t_comp == MATCH_GREATER_THAN)
+        return t > tv->t_value;
+    else /* tv->t_comp == MATCH_LESS_THAN */
+        return t < tv->t_value;
+}
+
+static struct list_constraint *create_timestamp_constraint (const char *type,
+                                                            json_t *values,
+                                                            flux_error_t *errp)
+{
+    struct timestamp_value *tv = NULL;
+    struct list_constraint *c;
+    json_t *v = json_array_get (values, 0);
+
+    if (!v) {
+        errprintf (errp, "timestamp value not specified");
+        return NULL;
+    }
+    if (json_is_string (v)) {
+        const char *str = json_string_value (v);
+        if (strstarts (str, ">="))
+            tv = timestamp_value_create_str (str + 2,
+                                             type,
+                                             MATCH_GREATER_THAN_EQUAL,
+                                             errp);
+        else if (strstarts (str, "<="))
+            tv = timestamp_value_create_str (str + 2,
+                                             type,
+                                             MATCH_LESS_THAN_EQUAL,
+                                             errp);
+        else if (strstarts (str, ">"))
+            tv = timestamp_value_create_str (str + 1,
+                                             type,
+                                             MATCH_GREATER_THAN,
+                                             errp);
+        else if (strstarts (str, "<"))
+            tv = timestamp_value_create_str (str + 1,
+                                             type,
+                                             MATCH_LESS_THAN,
+                                             errp);
+        else /* if no operation specified >= is default */
```

Comment:
I realize there isn't much point in supporting "=" for timestamps, but it feels weird that `>=` is the default here.
Maybe just support '=' and have that be the default, knowing that in most real use cases a bare timestamp value or `=` would never actually be used?

Reply:
Yeah, I picked `>=` since I figured it was the "best" of all the choices.

Defaulting to `=` does seem like an interesting idea and would make things consistent to everything else in the constraints.  However, I guess my concern would be if someone did a constraint like  `{"t_inactive": [ "SOMETHING" ]}`, basically this constraint would never return anything and a user could be confused.  I dunno if that's better or worse.

I almost feel that just requiring ">", "<", ">=", or "<=" would be better.  What do you think of just making it a requirement?

Reply:
I could go either way. I like making it a requirement better than having a default of `>=`. I don't personally think users are going to be constructing these constraints by hand, so I don't think allowing `{"t_inactive": [ "SOMETHING" ]}` would be confusing per se (but it would be difficult to test).